File: src/4-analyze-data.py
# 4-analyze-data.py
# ===========================================================================
# Regression analysis and table output for the earnings event study.
#
# This script produces:
#   1. Sample selection table (LaTeX via great_tables)
#   2. Frequency table by decade (LaTeX via great_tables)
#   3. Descriptive statistics (LaTeX via great_tables)
#   4. Correlation matrix (LaTeX via great_tables)
#   5. Regression table with SUE x SameSign interaction (LaTeX via pyfixest)
#
# Tables 1-4 use great_tables (by Posit), which provides a clean API for
# building publication-ready tables and exporting to LaTeX. Table 5 uses
# pyfixest's native etable(type="tex") because great_tables' LaTeX export
# does not yet support row stubs/groups, which pyfixest uses for the
# regression layout.
#
# HOW TO RUN:
#   uv run src/4-analyze-data.py
# ===========================================================================


# Setup ------------------------------------------------------------------------

import os
import sys

import great_tables as gt
import polars as pl
import pandas as pd
import pyfixest as pf
from dotenv import load_dotenv

sys.stdout.reconfigure(encoding="utf-8")

load_dotenv(".env", override=True)
data_dir = os.getenv("DATA_DIR")
output_dir = os.getenv("OUTPUT_DIR")

os.makedirs(output_dir, exist_ok=True)


# Read in the data from the previous step --------------------------------------

regdata = pl.read_parquet(f"{data_dir}/regdata.parquet")

# Standardize continuous controls for the interaction model.
regdata = regdata.with_columns(
    ((pl.col("log_mve") - pl.col("log_mve").mean())
     / pl.col("log_mve").std()).alias("log_mve_std")
)

print(f"regdata: {regdata.shape[0]:,} rows")


# =============================================================================
# Table 1: Sample Selection
# =============================================================================

# Read the step-by-step observation counts produced by script 2.
sample_sel = pl.read_parquet(f"{data_dir}/sample-selection.parquet")

# Compute the (Diff) column
sel_pd = sample_sel.to_pandas()
sel_pd["diff"] = sel_pd["obs"].shift(1) - sel_pd["obs"]
sel_pd["(Diff)"] = sel_pd["diff"].apply(
    lambda x: f"({x:,.0f})" if pd.notna(x) else ""
)
sel_pd["Obs"] = sel_pd["obs"].apply(lambda x: f"{x:,.0f}")

tbl = (gt.GT(sel_pd[["step", "description", "Obs", "(Diff)"]])
    .cols_label(step="Step", description="Description")
)

with open(f"{output_dir}/sample-selection-py.tex", "w", encoding="utf-8") as f:
    f.write(tbl.as_latex())
print("Table 1 (sample selection) saved")


# =============================================================================
# Table 2: Frequency by Decade
# =============================================================================

decade_data = (regdata
    .with_columns(
        pl.when(pl.col("fyearq").is_between(1970, 1979)).then(pl.lit("1970 - 1979"))
          .when(pl.col("fyearq").is_between(1980, 1989)).then(pl.lit("1980 - 1989"))
          .when(pl.col("fyearq").is_between(1990, 1999)).then(pl.lit("1990 - 1999"))
          .when(pl.col("fyearq").is_between(2000, 2009)).then(pl.lit("2000 - 2009"))
          .when(pl.col("fyearq").is_between(2010, 2019)).then(pl.lit("2010 - 2019"))
          .when(pl.col("fyearq") >= 2020).then(pl.lit("2020+"))
          .alias("Year")
    )
    .group_by("Year")
    .agg(
        pl.len().alias("Firm-Quarters"),
        pl.col("same_sign").sum().alias("SameSign Quarters"),
        pl.col("same_sign").mean().alias("Pct. SameSign"),
    )
    .sort("Year")
)

# Add total row
total = pl.DataFrame(
    {"Year": "Total",
     "Firm-Quarters": regdata.shape[0],
     "SameSign Quarters": int(regdata["same_sign"].sum()),
     "Pct. SameSign": float(regdata["same_sign"].mean())},
    schema=decade_data.schema,
)
decade_data = pl.concat([decade_data, total])

tbl = (gt.GT(decade_data.to_pandas())
    .fmt_integer(columns="Firm-Quarters", use_seps=True)
    .fmt_integer(columns="SameSign Quarters", use_seps=True)
    .fmt_percent(columns="Pct. SameSign", decimals=2)
)

with open(f"{output_dir}/freqtable-py.tex", "w", encoding="utf-8") as f:
    f.write(tbl.as_latex())
print("Table 2 (frequency by decade) saved")


# =============================================================================
# Table 3: Descriptive Statistics
# =============================================================================

descrip_vars = ["bhar", "sue", "same_sign", "loss", "log_mve"]
descrip_labels = {
    "bhar": "BHAR[-1,+1]",
    "sue": "SUE",
    "same_sign": "SameSign",
    "loss": "LOSS",
    "log_mve": "ln(MVE)",
}

stats = []
for var in descrip_vars:
    col = regdata[var].drop_nulls()
    stats.append({
        "Variable": descrip_labels[var],
        "N": col.len(),
        "Mean": col.mean(),
        "SD": col.std(),
        "Min": col.min(),
        "P25": col.quantile(0.25),
        "Median": col.median(),
        "P75": col.quantile(0.75),
        "Max": col.max(),
    })

descrip_df = pd.DataFrame(stats)

tbl = (gt.GT(descrip_df)
    .fmt_integer(columns="N", use_seps=True)
    .fmt_number(columns=["Mean", "SD", "Min", "P25", "Median", "P75", "Max"], decimals=3)
)

with open(f"{output_dir}/descrip-py.tex", "w", encoding="utf-8") as f:
    f.write(tbl.as_latex())
print("Table 3 (descriptive stats) saved")


# =============================================================================
# Table 4: Correlation Matrix
# =============================================================================

corrdata = (regdata
    .select(
        pl.col("bhar").alias("BHAR"),
        pl.col("sue").alias("SUE"),
        pl.col("same_sign").cast(pl.Float64).alias("SameSign"),
        pl.col("loss").cast(pl.Float64).alias("LOSS"),
        pl.col("log_mve").alias("SIZE"),
    )
    .to_pandas()
)

corr = corrdata.corr()
corr.insert(0, " ", corr.index)

tbl = (gt.GT(corr)
    .fmt_number(columns=["BHAR", "SUE", "SameSign", "LOSS", "SIZE"], decimals=3)
    .cols_label(**{" ": ""})
)

with open(f"{output_dir}/corrtable-py.tex", "w", encoding="utf-8") as f:
    f.write(tbl.as_latex())
print("Table 4 (correlation matrix) saved")


# =============================================================================
# Table 5: Regression Table
# =============================================================================

# pyfixest uses the same formula syntax as R's fixest:
#   "y ~ x1 * x2 | fe1 + fe2"
# Two-way clustering: vcov={"CRV1": "permno+fyearq"}

reg_pd = regdata.to_pandas()

m1 = pf.feols("bhar ~ sue", data=reg_pd)
m2 = pf.feols("bhar ~ sue * same_sign", data=reg_pd)
m3 = pf.feols("bhar ~ sue * same_sign | fyearq", data=reg_pd)
m4 = pf.feols("bhar ~ sue * same_sign | fyearq + ff12num", data=reg_pd)
m5 = pf.feols("bhar ~ sue * same_sign + sue * log_mve_std + sue * loss | fyearq + ff12num",
               data=reg_pd)

for m in [m1, m2, m3, m4, m5]:
    m.vcov({"CRV1": "permno+fyearq"})

# Generate LaTeX via pyfixest's native etable(type="tex").
# We use etable(type="tex") rather than etable(type="gt").as_latex() because
# great_tables' LaTeX export does not yet support row stubs/groups, which
# pyfixest uses to organize coefficients, FE indicators, and fit statistics.
# coef_fmt controls the display: b* adds significance stars to the coefficient,
# t:.2f formats t-statistics to 2 decimal places.
# felabels renames the raw FE variable names to plain English.
# custom_model_stats adds R² Within and Controls indicator rows.
# Labels use LaTeX math mode for variable names.
tex = pf.etable(
    [m1, m2, m3, m4, m5],
    type="tex",
    coef_fmt="b* \n (t:.2f)",
    signif_code=[0.01, 0.05, 0.1],
    keep=["sue$", "^same_sign$", "^sue:same_sign$"],
    labels={
        "sue": "$SUE$",
        "same_sign": "$SameSign$",
        "sue:same_sign": "$SUE \\times SameSign$",
    },
    # pyfixest uses slightly different internal FE names when FEs are absorbed
    # alone vs. together (e.g., "fyearq" vs "fyearq "). Map all variants.
    felabels={
        "fyearq": "Year FE",
        "fyearq ": "Year FE",
        "ff12num": "Industry FE",
        " ff12num": "Industry FE",
    },
    # Use show_fe=False and handle FE indicators via custom_model_stats,
    # because pyfixest's auto FE display duplicates "Year FE" when fyearq
    # appears in different absorb configurations across models.
    show_fe=False,
    custom_model_stats={
        "Year FE": ["", "", "Yes", "Yes", "Yes"],
        "Industry FE": ["", "", "", "Yes", "Yes"],
        "Controls": ["", "", "", "", "Yes"],
        "$R^2$ Within": [
            f"{m._r2_within:.3f}" if (m._r2_within is not None
                                      and str(m._r2_within) != "nan") else ""
            for m in [m1, m2, m3, m4, m5]
        ],
    },
    model_heads=["Base", "Interaction", "Year FE", "Two-Way FE", "Controls"],
)

with open(f"{output_dir}/regression-py.tex", "w", encoding="utf-8") as f:
    f.write(tex)
print("Table 5 (regression) saved")

# Print regression to console for review (plain text labels for markdown)
pf.etable(
    [m1, m2, m3, m4, m5],
    type="md",
    coef_fmt="b* \n (t:.2f)",
    signif_code=[0.01, 0.05, 0.1],
    keep=["sue$", "^same_sign$", "^sue:same_sign$"],
    labels={
        "sue": "SUE",
        "same_sign": "SameSign",
        "sue:same_sign": "SUE x SameSign",
    },
    felabels={
        "fyearq": "Year FE",
        "fyearq ": "Year FE",
        "ff12num": "Industry FE",
        " ff12num": "Industry FE",
    },
    show_fe=False,
    custom_model_stats={
        "Year FE": ["", "", "Yes", "Yes", "Yes"],
        "Industry FE": ["", "", "", "Yes", "Yes"],
        "Controls": ["", "", "", "", "Yes"],
        "R² Within": [
            f"{m._r2_within:.3f}" if (m._r2_within is not None
                                      and str(m._r2_within) != "nan") else ""
            for m in [m1, m2, m3, m4, m5]
        ],
    },
    model_heads=["Base", "Interaction", "Year FE", "Two-Way FE", "Controls"],
)

print(f"\nAll tables saved to: {output_dir}")

File: src/4-analyze-data.R
# 4-analyze-data.R
# ===========================================================================
# Regression analysis and table output for the earnings event study.
#
# This script produces all tables in BOTH LaTeX and Word formats from a
# single run. Models are fit once; each table is then exported twice:
#   - LaTeX .tex files (for Overleaf / pdflatex)
#   - A combined Word .docx file (via flextable + officer)
#
# Tables:
#   1. Sample selection (step-by-step obs counts from script 2)
#   2. Frequency table by decade (SameSign counts)
#   3. Descriptive statistics (modelsummary's datasummary)
#   4. Correlation matrix (Pearson above, Spearman below)
#   5. Regression table (SUE × SameSign interaction, FE, clustered SEs)
#
# HOW TO RUN:
#   Open in RStudio and run interactively, or:
#   Rscript src/4-analyze-data.R
# ===========================================================================


# Setup ------------------------------------------------------------------------

if (!require("pacman")) install.packages("pacman")
pacman::p_load(dotenv, modelsummary, sjlabelled, kableExtra, tinytable,
               formattable, flextable, officer, lubridate, glue, arrow, fixest,
               tidyverse)

load_dot_env(".env")
data_dir   <- Sys.getenv("DATA_DIR")
output_dir <- Sys.getenv("OUTPUT_DIR")

source("src/utils.R")

# Prevent modelsummary from adding extra formatting to LaTeX numbers
options(modelsummary_format_numeric_latex = "plain")


# Read in the data from the previous step --------------------------------------

regdata <- read_parquet(glue("{data_dir}/regdata.parquet")) |>
  # Standardize continuous controls for the interaction model. In a model with
  # SUE × control interactions, the SUE main effect is the ERC evaluated at
  # control = 0. Standardizing (mean 0, sd 1) makes that "at the sample mean."
  # standardize() is from utils.R: (x - mean) / sd. We use it instead of
  # base::scale() because scale() returns a matrix, not a vector.
  mutate(log_mve_std = standardize(log_mve)) |>
  sjlabelled::var_labels(
    bhar        = "$BHAR_{[-1,+1]}$",
    sue         = "$SUE$",
    same_sign   = "$SameSign$",
    loss        = "$LOSS$",
    log_mve     = "$\\ln(MVE)$",
    log_mve_std = "$\\ln(MVE)^{std}$"
  )

nrow(regdata)


# =============================================================================
# Table 1: Sample Selection
# =============================================================================

sample_selection <- read_parquet(glue("{data_dir}/sample-selection.parquet"))

sample_table <- sample_selection |>
  mutate(
    diff = dplyr::lag(obs) - obs,
    Obs  = formattable::comma(obs, digits = 0),
    `(Diff)` = if_else(is.na(diff), "",
                       paste0("(", formattable::comma(diff, digits = 0), ")"))
  ) |>
  transmute(Step = as.character(step),
            Description = description,
            Obs,
            `(Diff)`)

sample_table

# --- LaTeX ---
kbl(sample_table, format = "latex", booktabs = TRUE,
    linesep = "", align = "clrr", escape = TRUE) |>
  save_kable(glue("{output_dir}/sample-selection-r.tex"))

# --- Word (flextable) ---
ft_sample <- flextable(sample_table) |>
  set_table_properties(layout = "autofit") |>
  fit_to_width(max_width = 6.5)


# =============================================================================
# Table 2: Observations by Decade
# =============================================================================

table2 <- regdata |>
  mutate(Year = case_when(
    fyearq %in% 1970:1979 ~ "1970 - 1979",
    fyearq %in% 1980:1989 ~ "1980 - 1989",
    fyearq %in% 1990:1999 ~ "1990 - 1999",
    fyearq %in% 2000:2009 ~ "2000 - 2009",
    fyearq %in% 2010:2019 ~ "2010 - 2019",
    fyearq >= 2020 ~ "2020+")
  ) |>
  group_by(Year) |>
  summarize(`Firm-Quarters`     = formattable::comma(n(), digits = 0),
            `SameSign Quarters` = formattable::comma(sum(same_sign, na.rm = TRUE), digits = 0),
            `Pct. SameSign`     = formattable::percent(mean(same_sign, na.rm = TRUE), digits = 2))

totalrow <- regdata |>
  summarize(`Firm-Quarters`     = formattable::comma(n(), digits = 0),
            `SameSign Quarters` = formattable::comma(sum(same_sign, na.rm = TRUE), digits = 0),
            `Pct. SameSign`     = formattable::percent(mean(same_sign, na.rm = TRUE), digits = 2)) |>
  mutate(Year = "Total")

table2 <- bind_rows(table2, totalrow)
table2

# --- LaTeX ---
kbl(table2, format = "latex", booktabs = TRUE, linesep = "") |>
  save_kable(glue("{output_dir}/freqtable-r.tex"))

# --- Word ---
ft_freq <- flextable(table2) |> autofit()


# =============================================================================
# Table 3: Descriptive Statistics
# =============================================================================

#' Default number formatter for descriptive-stat tables: 3-digit comma format.
#' Used as the `fmt` argument to `datasummary()` so all numeric stats
#' (mean, SD, min, max, etc.) print with the same precision and
#' thousands separators.
my_fmt <- function(x) formattable::comma(x, digits = 3)

#' Non-missing-count formatter for descriptive-stat tables.
#' Returns the count as an integer with thousands separators, formatted
#' to slot into a `datasummary()` `(N = NN)` column. Special-case: if
#' the column is a logical of all-NA (a `datasummary` quirk for
#' character columns), report the column length instead.
NN <- function(x) {
  out <- if (is.logical(x) && all(is.na(x))) length(x) else sum(!is.na(x))
  formattable::comma(out, digits = 0)
}

descripdata <- regdata |>
  select(bhar, sue, same_sign, loss, log_mve) |>
  label_to_colnames()

# --- LaTeX ---
datasummary(All(descripdata) ~ (N = NN) +
              (Mean + SD + Min + P25 + Median + P75 + Max) * Arguments(fmt = my_fmt),
            data = descripdata,
            escape = FALSE,
            output = glue("{output_dir}/descrip-r.tex"))

# --- Word ---
ft_descrip <- datasummary(All(descripdata) ~ (N = NN) +
                            Mean * Arguments(fmt = my_fmt) +
                            SD * Arguments(fmt = my_fmt) +
                            Min * Arguments(fmt = my_fmt) +
                            P25 * Arguments(fmt = my_fmt) +
                            Median * Arguments(fmt = my_fmt) +
                            P75 * Arguments(fmt = my_fmt) +
                            Max * Arguments(fmt = my_fmt),
                          escape = FALSE,
                          output = 'flextable',
                          data = descripdata) |>
  fit_to_width(max_width = 6.5)


# =============================================================================
# Table 4: Correlation Matrix
# =============================================================================

# --- LaTeX ---
datasummary_correlation(descripdata,
                        method = "pearspear",
                        output = glue("{output_dir}/corrtable-r.tex"),
                        escape = FALSE)

# --- Word ---
ft_corr <- datasummary_correlation(descripdata,
                                    method = "pearspear",
                                    output = "flextable") |>
  fit_to_width(max_width = 6.5)


# =============================================================================
# Table 5: Regression Table
# =============================================================================

# Fit all 5 model specifications (done once, used for both LaTeX and Word)
models <- list(
  "Base" = feols(bhar ~ sue,
                 regdata, fixef.rm = "singletons"),
  "Interaction" = feols(bhar ~ sue * same_sign,
                        regdata, fixef.rm = "singletons"),
  "Year FE" = feols(bhar ~ sue * same_sign | fyearq,
                     regdata, fixef.rm = "singletons"),
  "Two-Way FE" = feols(bhar ~ sue * same_sign | fyearq + ff12num,
                        regdata, fixef.rm = "singletons"),
  "Controls" = feols(bhar ~ sue * same_sign + sue * log_mve_std + sue * loss |
                       fyearq + ff12num,
                     regdata, fixef.rm = "singletons")
)

#' S3 method that adds a "Controls" indicator row to modelsummary tables.
#'
#' `modelsummary` calls `glance_custom()` on each model to gather extra
#' goodness-of-fit columns. Defining `glance_custom.fixest` (the
#' fixest-specific S3 method) lets us inject a custom row that says "X"
#' when the model's coefficients include our control set and is blank
#' otherwise. The rest of the gof formatting is configured via `gm`
#' below.
#'
#' @param x A fitted `fixest` model (typically from `feols()`).
#' @param ... Ignored; required for the S3 method signature.
#' @return One-row data frame with column `Controls`, value `"X"` or `""`.
glance_custom.fixest <- function(x, ...) {
  controls <- c("log_mve_std", "loss")
  if (all(controls %in% names(coef(x)))) {
    data.frame(Controls = "X")
  } else {
    data.frame(Controls = "")
  }
}

# Goodness-of-fit statistics
gm <- list(
  list("raw" = "FE: fyearq",  "clean" = "Year FE",      "fmt" = NULL),
  list("raw" = "FE: ff12num", "clean" = "Industry FE",   "fmt" = NULL),
  list("raw" = "Controls",    "clean" = "Controls",      "fmt" = NULL),
  list("raw" = "nobs",        "clean" = "N",             "fmt" = function(x) formattable::comma(x, digits = 0)),
  list("raw" = "r.squared",   "clean" = "$R^2$",         "fmt" = 3),
  list("raw" = "r2.within",   "clean" = "$R^2$ Within",  "fmt" = 3)
)

# --- LaTeX ---
# Coefficient map with LaTeX math notation
cm_latex <- c(
  "sue"            = "$SUE$",
  "same_sign"      = "$SameSign$",
  "sue:same_sign"  = "$SUE \\times SameSign$"
)

modelsummary(models,
             vcov = ~ permno + fyearq,
             statistic = "statistic",
             stars = c('\\sym{*}' = .1, '\\sym{**}' = .05, '\\sym{***}' = .01),
             estimate = "{estimate}{stars}",
             coef_map = cm_latex,
             gof_map = gm,
             output = glue("{output_dir}/regression-r.tex"),
             escape = FALSE)

# --- Word ---
# Coefficient map with plain text labels
cm_word <- c(
  "sue"            = "SUE",
  "same_sign"      = "SameSign",
  "sue:same_sign"  = "SUE x SameSign"
)

# Update gof_map: remove LaTeX math from R² labels for Word
gm_word <- gm
gm_word[[5]]$clean <- "R-squared"
gm_word[[6]]$clean <- "R-sq Within"

ft_reg <- modelsummary(models,
                       vcov = ~ permno + fyearq,
                       statistic = "statistic",
                       stars = c('*' = .1, '**' = .05, '***' = .01),
                       estimate = "{estimate}{stars}",
                       coef_map = cm_word,
                       gof_map = gm_word,
                       output = "flextable") |>
  fit_to_width(max_width = 6.5) |>
  autofit()

# Preview
ft_reg


# =============================================================================
# Assemble Word document
# =============================================================================

# The officer package builds Word documents programmatically: headings,
# paragraphs, tables (flextables), images, and page breaks.

read_docx() |>
  body_add_par("Sample Selection", style = "heading 1") |>
  body_add_par("") |>
  body_add_flextable(value = ft_sample) |>
  body_add_break(pos = "after") |>
  body_add_par("Sample Frequency", style = "heading 1") |>
  body_add_par("") |>
  body_add_flextable(value = ft_freq) |>
  body_add_break(pos = "after") |>
  body_add_par("Descriptive Statistics", style = "heading 1") |>
  body_add_par("") |>
  body_add_flextable(value = ft_descrip) |>
  body_add_break(pos = "after") |>
  body_add_par("Correlation Matrix", style = "heading 1") |>
  body_add_par("") |>
  body_add_flextable(value = ft_corr) |>
  body_add_break(pos = "after") |>
  body_add_par("Regression Table", style = "heading 1") |>
  body_add_par("") |>
  body_add_flextable(value = ft_reg) |>
  body_add_break(pos = "after") |>
  body_add_par("Figure: SameSign by Industry", style = "heading 1") |>
  body_add_par("") |>
  body_add_img(glue("{output_dir}/ff12_fig.png"),
               height = 3.6, width = 4.2, style = "centered") |>
  body_add_break(pos = "after") |>
  body_add_par("Figure: Event Study CAR", style = "heading 1") |>
  body_add_par("") |>
  body_add_img(glue("{output_dir}/car_fig.png"),
               height = 3.6, width = 4.2, style = "centered") |>
  print(target = glue("{output_dir}/tables-r.docx"))

cat("All tables saved to:", output_dir, "\n")
cat("  LaTeX: sample-selection-r.tex, freqtable-r.tex, descrip-r.tex,\n")
cat("         corrtable-r.tex, regression-r.tex\n")
cat("  Word:  tables-r.docx\n")
